components/script/dom/htmlprogresselement.rs

```diff
@@ -48,4 +54,56 @@ impl HTMLProgressElement {
 impl HTMLProgressElementMethods for HTMLProgressElement {
     // https://html.spec.whatwg.org/multipage/#dom-lfe-labels
     make_labels_getter!(Labels, labels_node_list);
+
+    // https://html.spec.whatwg.org/multipage/#attr-progress-value
+    fn Value(&self) -> Finite<f64> {
+        let value = self.value.get().unwrap_or(Finite::wrap(0.0));
+        if *value > *self.max.get() {
+            self.max.get()
+        } else {
+            value
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#attr-progress-value
+    fn SetValue(&self, new_val: Finite<f64>) {
+        if 0.0 <= *new_val {
+            self.value.set(Some(new_val))
+        }
+    }
+
+    // https://html.spec.whatwg.org/multipage/#attr-progress-max
+    fn Max(&self) -> Finite<f64> {
+        self.max.get()
+    }
+
+    // https://html.spec.whatwg.org/multipage/#attr-progress-max
+    fn SetMax(&self, new_val: Finite<f64>) {
+        if *new_val > 0.0 {
+            self.max.replace(new_val);
+        }
```

Comment:
Ah, I see two problems from reading the spec. The `max` IDL attribute is supposed to reflect the content attribute (so doing `progress.setAttribute('max', 5.0)` then `progress.max` should return 5.0), but that doesn't happen right now. To fix this, we should not store the `max` value as a field in the struct; we should always use `self.upcast::<Element>().get_string_attribute(local_name!("max")).parse()` and handle the result appropriately when we need to calculate the maximum according to https://html.spec.whatwg.org/multipage/form-elements.html#concept-progress-maximum.

Similarly, rather than storing the `value` value as a field, we should do the same thing when we need to calculate a current value according to https://html.spec.whatwg.org/multipage/form-elements.html#concept-progress-value.

Reply:
Thanks a lot, and sorry for my inexperience